<a href="https://colab.research.google.com/github/jagdegun/AM1/blob/main/L6_AM1_Jagjit_Degun.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
# Basic libraries
import numpy as np
import pandas as pd

# Visualization (optional but helpful)
import matplotlib.pyplot as plt
import seaborn as sns

# Sklearn tools
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import classification_report, confusion_matrix, accuracy_score

In [ ]:
df = pd.read_excel('/content/Level 6 AM1 Dataset.xlsx')
display(df)

,Start Date,End Date,Consultation,Location,Postcode,Gender,Age,Advertisement Type,Advertisement Release Time
0,2025-08-26 08:45:59,2025-08-26 08:50:36,Northampton Town Centre,Northampton,NN12 8XA,Female,45 to 54,Council newsletter or email,2025-07-15 12:10:00
1,2025-08-25 20:31:23,2025-08-25 23:34:17,Northampton Town Centre,Northampton,N4 2FG,Male,45 to 54,Facebook,2025-07-15 12:10:00
2,2025-08-25 22:57:05,2025-08-25 23:17:07,Northampton Town Centre,Northampton,nn15nb,Male,65 to 74,Other (please specify),2025-07-15 12:10:00
3,2025-08-25 22:36:45,2025-08-25 22:42:49,Northampton Town Centre,Northampton,Nn7 1du,Female,65 to 74,Facebook,2025-07-15 12:10:00
4,2025-08-25 20:34:21,2025-08-25 20:45:21,Northampton Town Centre,Northampton,NN1 4QS,Female,35 to 44,Word of mouth,2025-07-15 12:10:00
...,...,...,...,...,...,...,...,...,...
2286,2024-09-11 15:07:59,2024-09-11 15:08:08,Project Thrive,Birmingham,NaN,NaN,NaN,NaN,NaT
2287,2024-09-10 17:41:07,2024-09-10 17:41:17,Project Thrive,Birmingham,NaN,NaN,NaN,NaN,NaT
2288,2024-09-10 10:31:29,2024-09-10 10:32:28,Project Thrive,Birmingham,b46 1sp,Female,35 to 44,Word of mouth,NaT
2289,2024-09-09 11:43:09,2024-09-09 11:43:28,Project Thrive,Birmingham,NaN,NaN,NaN,NaN,NaT


In [ ]:
y = df['Advertisement Type']

In [ ]:
features = [
    'Consultation',
    'Location',
    'Gender',
    'Age',
    'Advertisement Release Time '
]


In [ ]:
print(df.columns)

Index(['Start Date', 'End Date', 'Consultation', 'Location', 'Postcode',
       'Gender', 'Age', 'Advertisement Type', 'Advertisement Release Time '],
      dtype='object')


In [ ]:
X = df[features]
y = df['Advertisement Type']

In [ ]:
X = pd.get_dummies(X, drop_first=True)

In [ ]:
X.head()

,Advertisement Release Time,Consultation_Northampton Town Centre,Consultation_Project Thrive,Location_Derby,Location_Northampton,Gender_Male,Gender_Other (please specify),Gender_Prefer not to say,Gender_Prefer to self-describe:,Age_25 to 34,Age_35 to 44,Age_45 to 54,Age_55 to 64,Age_65 to 74,Age_75+,Age_Under 18
0,2025-07-15 12:10:00,True,False,False,True,False,False,False,False,False,False,True,False,False,False,False
1,2025-07-15 12:10:00,True,False,False,True,True,False,False,False,False,False,True,False,False,False,False
2,2025-07-15 12:10:00,True,False,False,True,True,False,False,False,False,False,False,False,True,False,False
3,2025-07-15 12:10:00,True,False,False,True,False,False,False,False,False,False,False,False,True,False,False
4,2025-07-15 12:10:00,True,False,False,True,False,False,False,False,False,True,False,False,False,False,False


In [ ]:
df.isnull().sum()

,0
Start Date,0
End Date,0
Consultation,0
Location,0
Postcode,33
Gender,33
Age,33
Advertisement Type,33
Advertisement Release Time,96


In [ ]:
df = df.dropna()

In [ ]:
X = df[features]
y = df['Advertisement Type']

X = pd.get_dummies(X, drop_first=True)


In [ ]:
X_train, X_test, y_train, y_test = train_test_split(
    X, y,
    test_size=0.2,
    random_state=42,
    stratify=y
)

In [ ]:
X.dtypes

,0
Advertisement Release Time,datetime64[ns]
Consultation_Northampton Town Centre,bool
Location_Northampton,bool
Gender_Male,bool
Gender_Other (please specify),bool
Gender_Prefer not to say,bool
Gender_Prefer to self-describe:,bool
Age_25 to 34,bool
Age_35 to 44,bool
Age_45 to 54,bool


In [ ]:
df['Advertisement Release Time '] = df['Advertisement Release Time '].astype(str)

In [ ]:
X = df[features]
X = pd.get_dummies(X, drop_first=True)

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(
    X, y,
    test_size=0.2,
    random_state=42,
    stratify=y
)

In [ ]:
scaler = StandardScaler()

X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)

In [ ]:
X.dtypes

,0
Consultation_Northampton Town Centre,bool
Location_Northampton,bool
Gender_Male,bool
Gender_Other (please specify),bool
Gender_Prefer not to say,bool
Gender_Prefer to self-describe:,bool
Age_25 to 34,bool
Age_35 to 44,bool
Age_45 to 54,bool
Age_55 to 64,bool


In [ ]:
model = LogisticRegression(
    multi_class='multinomial',
    max_iter=1000,
    solver='lbfgs'
)

model.fit(X_train, y_train)


/usr/local/lib/python3.12/dist-packages/sklearn/linear_model/_logistic.py:1247: FutureWarning: 'multi_class' was deprecated in version 1.5 and will be removed in 1.7. From then on, it will always use 'multinomial'. Leave it to its default value to avoid this warning.
  warnings.warn(


LogisticRegression(max_iter=1000, multi_class='multinomial')

In [ ]:
y_pred = model.predict(X_test)

In [ ]:
print("Accuracy:", accuracy_score(y_test, y_pred))
print("\nConfusion Matrix:\n", confusion_matrix(y_test, y_pred))
print("\nClassification Report:\n", classification_report(y_test, y_pred))

Accuracy: 0.45785876993166286

Confusion Matrix:
 [[  0  40   0   0   0   2   0   0   0]
 [  0 149   0   0   0   4   0   0   0]
 [  0  29   0   0   0   2   0   0   0]
 [  0   0   0   6   0   0   0  16   0]
 [  0  67   0   0   0   4   0   1   0]
 [  0  22   0   2   0   7   0   9   0]
 [  0   1   0   0   0   0   0   0   0]
 [  0   0   0   2   0   2   0  39   0]
 [  0  22   0   2   0   3   0   8   0]]

Classification Report:
                              precision    recall  f1-score   support

Council newsletter or email       0.00      0.00      0.00        42
                   Facebook       0.45      0.97      0.62       153
                  Instagram       0.00      0.00      0.00        31
                    Leaflet       0.50      0.27      0.35        22
               News article       0.00      0.00      0.00        72
     Other (please specify)       0.29      0.17      0.22        40
                      Radio       0.00      0.00      0.00         1
               Socia

/usr/local/lib/python3.12/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.12/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.12/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))


In [ ]:
from sklearn.linear_model import LogisticRegression

# Multiclass logistic regression with class balancing
model = LogisticRegression(
    multi_class='multinomial',
    max_iter=1000,
    solver='lbfgs',
    class_weight='balanced'  # <-- this tells the model to handle class imbalance
)

model.fit(X_train, y_train)
y_pred = model.predict(X_test)


/usr/local/lib/python3.12/dist-packages/sklearn/linear_model/_logistic.py:1247: FutureWarning: 'multi_class' was deprecated in version 1.5 and will be removed in 1.7. From then on, it will always use 'multinomial'. Leave it to its default value to avoid this warning.
  warnings.warn(


In [ ]:
from sklearn.metrics import classification_report, confusion_matrix, accuracy_score

print("Accuracy:", accuracy_score(y_test, y_pred))
print("\nConfusion Matrix:\n", confusion_matrix(y_test, y_pred))
print("\nClassification Report:\n", classification_report(y_test, y_pred))

Accuracy: 0.214123006833713

Confusion Matrix:
 [[ 6  8 11  0  2  1 13  0  1]
 [32 20 29  0  1  3 50  0 18]
 [ 5  4 13  0  0  1  7  0  1]
 [ 0  0  0 13  0  0  0  9  0]
 [ 7  6 17  0  0  3 33  1  5]
 [ 1  2 10  4  0  4  9  7  3]
 [ 0  0  0  0  0  0  1  0  0]
 [ 0  0  0  9  0  2  0 32  0]
 [ 3  2  5  3  1  2  7  7  5]]

Classification Report:
                              precision    recall  f1-score   support

Council newsletter or email       0.11      0.14      0.12        42
                   Facebook       0.48      0.13      0.21       153
                  Instagram       0.15      0.42      0.22        31
                    Leaflet       0.45      0.59      0.51        22
               News article       0.00      0.00      0.00        72
     Other (please specify)       0.25      0.10      0.14        40
                      Radio       0.01      1.00      0.02         1
               Social media       0.57      0.74      0.65        43
              Word of mouth       

In [ ]:
import numpy as np

cm = confusion_matrix(y_test, y_pred)
pred_counts = cm.sum(axis=0)
pred_percent = pred_counts / pred_counts.sum()

print(pred_percent)

[0.12300683 0.09567198 0.19362187 0.06605923 0.00911162 0.03644647
 0.27334852 0.12756264 0.07517084]


In [ ]:
import pandas as pd

new_data = pd.DataFrame([{
    'Consultation': 'Online',
    'Location': 'Birmingham',
    'Gender': 'Female',
    'Age': 28,
    'Advertisement Release Time': 'Morning'
}])

In [ ]:
X = pd.get_dummies(X, drop_first=True)

### Project Summary: Predicting Advertisement Effectiveness via Demographic Analysis

**Paragraph 1: Problem Area and Business Benefit**  
The project addresses the challenge of optimizing marketing reach for public consultations. Organizations often struggle to identify which advertisement channels (e.g., Facebook, Leaflets, Social Media) effectively engage specific demographic groups. By analyzing the 'Level 6 AM1 Dataset,' the business aims to understand the relationship between a participant's location, gender, age, and how they discovered the consultation. Addressing this problem is beneficial because it allows for more targeted marketing spend, reducing costs while increasing public participation and ensuring that diverse community voices are reached more efficiently.

**Paragraph 2: The Solution and ML Technique**  
My solution involved building a multiclass classification model to predict the 'Advertisement Type' based on features such as Consultation type, Location, Gender, and Age. I utilized **Multinomial Logistic Regression** as the core machine learning technique. To prepare the data, I handled missing values through row deletion and transformed categorical variables using one-hot encoding. Initial results showed a strong bias toward majority classes (like Facebook), so I refined the solution by implementing a **'balanced' class weight** strategy within the Logistic Regression framework and utilized **StandardScaler** to normalize the feature space. This ensured the model accounted for less frequent but equally important advertisement channels.

**Paragraph 3: Outcomes and Reception**  
Following the implementation of class balancing, the model's macro-average recall improved, providing a more equitable look at how different groups interact with various media. While the overall accuracy decreased (to approximately 21%), the model became significantly better at identifying minority classes like 'Radio' and 'Leaflets' which were previously ignored. This work was received as a valuable exploratory tool; it highlighted significant data imbalances and identified which demographics are currently underserved by digital-only marketing. The insights provided a data-driven foundation for shifting towards a multi-channel advertisement strategy to improve inclusivity in future consultations.